pkg/operator/ceph/cluster/osd/health.go

```diff
@@ -69,16 +69,16 @@ func NewOSDHealthMonitor(context *clusterd.Context, clusterInfo *client.ClusterI
 }
 
 // Start runs monitoring logic for osds status at set intervals
-func (m *OSDHealthMonitor) Start(stopCh chan struct{}) {
+func (m *OSDHealthMonitor) Start(context context.Context) {
 
 	for {
 		select {
 		case <-time.After(*m.interval):
 			logger.Debug("checking osd processes status.")
 			m.checkOSDHealth()
 
-		case <-stopCh:
-			logger.Infof("Stopping monitoring of OSDs in namespace %q", m.clusterInfo.Namespace)
+		case <-context.Done():
+			logger.Infof("stopping monitoring of OSDs in namespace %q", m.clusterInfo.Namespace)
 			return
 		}
 	}
```

Comment:
@leseb This block makes sense to me and matches my understanding of how contexts are intended to work. We have the health checker goroutine, and it has a copy of the context. When the context is canceled, this goroutine will gracefully exit once it has finished its current health check operation. I also see this for the mon health checker and object store checker.

Does this context get canceled each time the CephCluster is updated and needs to be re-reconciled? Or is it only when the operator is being shut down?

In the high-level controller, when the context is canceled, does the code verify that all health checker goroutines are stopped before starting operation again? Or is this even a concern?

I don't see any code code like this in the code that reconciles OSDs. If Rook is partway through creating and updating a bunch of OSDs, how does Rook stop that operation so quickly when it never checks for `context.Done`? To my understanding, it is because of the SIGHUP signal which causes the go program to exit.

Go documentation of Cancel function (which I assume work the same way as with our controller manager) suggest to me that Cancel() merely closes the "Done" channel which signals goroutines should gracefully stop. But it doesn't forcefully cancel the goroutine. https://pkg.go.dev/context#CancelFunc 

AFAICT, there is no way to forcefully terminate a goroutine except to gracefully handle the channel ([example](https://stackoverflow.com/a/6855549)). I am therefore confused about how we are stopping a CephCluster reconcile partway through its operation unless the SIGHUP signal is exiting the program and goroutines via Go's internal panic mechanism [as documented](https://pkg.go.dev/os/signal#hdr-Default_behavior_of_signals_in_Go_programs)

Reply:
I was looking at this file: `pkg/operator/ceph/operator.go`:

In it, I see that we call a `CancelFunc`, but I don't see that we are protecting ourselves from running multiple reconciles at once. Docs for `CancelFunc` are below.
> A CancelFunc tells an operation to abandon its work. A CancelFunc does not wait for the work to stop. A CancelFunc may be called by multiple goroutines simultaneously. After the first call, subsequent calls to a CancelFunc do nothing.

Reply:
> @leseb This block makes sense to me and matches my understanding of how contexts are intended to work. We have the health checker goroutine, and it has a copy of the context. When the context is canceled, this goroutine will gracefully exit once it has finished its current health check operation. I also see this for the mon health checker and object store checker.
> 
> Does this context get canceled each time the CephCluster is updated and needs to be re-reconciled? Or is it only when the operator is being shut down?

This happens each time the CephCluster is updated and needs to be re-reconciled so during that small window the health status will not be updated.

> 
> In the high-level controller, when the context is canceled, does the code verify that all health checker goroutines are stopped before starting operation again? Or is this even a concern?

The code does not verify this and honestly, I have no idea how we could potentially validate that, despite looking at the logs. Typically when the context is canceled the operator logs will print a line like: `stopping monitoring of ceph status`.

> 
> I don't see any code code like this in the code that reconciles OSDs. If Rook is partway through creating and updating a bunch of OSDs, how does Rook stop that operation so quickly when it never checks for `context.Done`? To my understanding, it is because of the SIGHUP signal which causes the go program to exit.

I'm answering this at the very end. SIGHUP signal only causes the go program to exit if not using a signal handler. We do use a signal handler with this code:

```go
// Used to watch for operator's config changes
configChan := make(chan os.Signal, 1)
signal.Notify(configChan, syscall.SIGHUP)
```

Then we wait for this channel to be populated with `case <-configChan:`. Then at this point, we can decide whatever we want to do as an action. In our case, we call `opManagerStop()` which is the cancel function of the main context.

> 
> Go documentation of Cancel function (which I assume work the same way as with our controller manager) suggest to me that Cancel() merely closes the "Done" channel which signals goroutines should gracefully stop. But it doesn't forcefully cancel the goroutine. https://pkg.go.dev/context#CancelFunc

We cannot forcefully stop a goroutine, this concept does not exist AFAIK, that's why we use channel or context (these days contexts are preferred). Where each routine at the very end of it 

> 
> AFAICT, there is no way to forcefully terminate a goroutine except to gracefully handle the channel ([example](https://stackoverflow.com/a/6855549)). I am therefore confused about how we are stopping a CephCluster reconcile partway through its operation unless the SIGHUP signal is exiting the program and goroutines via Go's internal panic mechanism [as documented](https://pkg.go.dev/os/signal#hdr-Default_behavior_of_signals_in_Go_programs)

Ok, let me recap here. The reason why the reconcile can be stopped so rapidly is because the context that gets passed down to the controller runtime manager is cancelled. This is how it starts:

```go
logger.Info("starting the controller-runtime manager")
if err := mgr.Start(context); err != nil {
	mgrErrorCh <- errors.Wrap(err, "failed to run the controller-runtime manager")
	return
}
```

Start starts all registered Controllers and blocks until the context is canceled.
Internally, the controller runtime manager code runs all the controllers in a separate goroutine, each controller as a worker number which represents the maximum number of concurrent Reconciles which can be run. Defaults to 1. [This](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L148) is the entrypoint of all that.  This is the code that manages runs and cancellations through the context.

As a quick experiment when turning on logging of the controller runtime, we can see that things are started and stopped correctly:
When started:
```log
{"level":"info","ts":1631176483.1089125,"logger":"controller-runtime.manager.controller.ceph-nfs-controller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1631176483.109101,"logger":"controller-runtime.manager.controller.rook-ceph-operator-bucket-controller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1631176483.1091754,"logger":"controller-runtime.manager.controller.ceph-object-controller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1631176483.1094518,"logger":"controller-runtime.manager.controller.ceph-client-controller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1631176483.1099894,"logger":"controller-runtime.manager.controller.ceph-cluster-controller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1631176483.1101332,"logger":"controller-runtime.manager.controller.ceph-rbd-mirror-controller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1631176483.1100574,"logger":"controller-runtime.manager.controller.ceph-file-controller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1631176483.1103122,"logger":"controller-runtime.manager.controller.ceph-object-realm-controller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1631176483.1100326,"logger":"controller-runtime.manager.controller.ceph-filesystem-mirror-controller","msg":"Starting workers","worker count":1}
```

When cancelled:

```log
2021-09-09 08:40:48.217710 I | operator: reloading operator's CRDs manager, cancelling all orchestrations!
{"level":"info","ts":1631176848.2178133,"logger":"controller-runtime.manager.controller.rook-ceph-operator-bucket-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2178204,"logger":"controller-runtime.manager.controller.rook-ceph-operator-csi-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2178855,"logger":"controller-runtime.manager.controller.ceph-object-zone-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2178998,"logger":"controller-runtime.manager.controller.rook-ceph-operator-config-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179046,"logger":"controller-runtime.manager.controller.ceph-object-zonegroup-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.217909,"logger":"controller-runtime.manager.controller.ceph-object-store-user-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179132,"logger":"controller-runtime.manager.controller.rook-ceph-operator-flex-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179174,"logger":"controller-runtime.manager.controller.ceph-block-pool-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179532,"logger":"controller-runtime.manager.controller.ceph-crashcollector-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.217962,"logger":"controller-runtime.manager.controller.clusterdisruption-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179668,"logger":"controller-runtime.manager.controller.ceph-nfs-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179773,"logger":"controller-runtime.manager.controller.ceph-filesystem-mirror-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179873,"logger":"controller-runtime.manager.controller.ceph-object-realm-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179923,"logger":"controller-runtime.manager.controller.ceph-file-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2179964,"logger":"controller-runtime.manager.controller.ceph-rbd-mirror-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.218001,"logger":"controller-runtime.manager.controller.ceph-cluster-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.218003,"logger":"controller-runtime.manager.controller.ceph-client-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.218017,"logger":"controller-runtime.manager.controller.ceph-object-controller","msg":"Shutdown signal received, waiting for all workers to finish"}
{"level":"info","ts":1631176848.2180326,"logger":"controller-runtime.webhook","msg":"shutting down webhook server"}
I0909 08:40:48.218076       7 manager.go:148] objectbucket.io/provisioner-manager "msg"="stopping provisioner"  "name"="rook-ceph.ceph.rook.io/bucket" "reason"={}
{"level":"info","ts":1631176848.2180734,"logger":"controller-runtime.manager.controller.ceph-client-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.21812,"logger":"controller-runtime.manager.controller.ceph-object-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.218125,"logger":"controller-runtime.manager.controller.ceph-object-realm-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181315,"logger":"controller-runtime.manager.controller.ceph-cluster-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181356,"logger":"controller-runtime.manager.controller.ceph-filesystem-mirror-controller","msg":"All workers finished"}
2021-09-09 08:40:48.218135 I | ceph-object-controller: deleting object "rookHealthCheckTestObject" from bucket "rook-ceph-bucket-checker-bd6bbd9e-39bb-495d-9cda-2c059e53f4a2" in object store "my-store"
{"level":"info","ts":1631176848.2181506,"logger":"controller-runtime.manager.controller.ceph-file-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181556,"logger":"controller-runtime.manager.controller.rook-ceph-operator-bucket-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.218163,"logger":"controller-runtime.manager.controller.clusterdisruption-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181664,"logger":"controller-runtime.manager.controller.rook-ceph-operator-config-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181702,"logger":"controller-runtime.manager.controller.rook-ceph-operator-flex-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181706,"logger":"controller-runtime.manager.controller.rook-ceph-operator-csi-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181828,"logger":"controller-runtime.manager.controller.ceph-rbd-mirror-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181888,"logger":"controller-runtime.manager.controller.ceph-object-zonegroup-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181897,"logger":"controller-runtime.manager.controller.ceph-nfs-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2181945,"logger":"controller-runtime.manager.controller.ceph-object-store-user-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.218199,"logger":"controller-runtime.manager.controller.ceph-crashcollector-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.2180803,"logger":"controller-runtime.manager.controller.ceph-object-zone-controller","msg":"All workers finished"}
{"level":"info","ts":1631176848.218192,"logger":"controller-runtime.manager.controller.ceph-block-pool-controller","msg":"All workers finished"}
```

Then the manager starts again.
Hope that clarifies.

Reply:
It does, thanks! It seems that the controller runtime manager actually does wait on the controllers to finish. I think this means that if an CephCluster update operation is in progress, controller runtime will wait on the reconcile to finish before proceeding. I will build the code and try it out today to see. But this could mean that OSD updates starve other CephCluster updates, which we don't want. I think we need to occasionally check if the context is canceled in our reconcile routine and gracefully exit still, just as we did before.

Reply:
If a reconcile is in progress and the context is cancelled then the reconcile stops immediately. But yes build it and try it :)

Reply:
I still don't understand how it's really happening, but it clearly is, and it's what we want, so I won't question it further. 😂 

Thanks for helping me through understanding (or trying to understand) what's going on here.

Reply:
I had another look at this in the controller runtime code and I was wrong. Each reconciles has a context and is part of a goroutine. However each of them is part of a WaitGroup, when cancel is called, the controller waits for all the workers to return before returning from the cancellation. This effectively means the reconciles will keep running and a new one will be trigger once the controller starts again.
We don't want that.

@BlaineEXE thanks for insisting on this. At the end of the day, there is no magic and any ongoing execution cannot be "killed", as goroutine cannot be stopped either if they deadlock. Just returning would mean that we leak goroutine on purpose too which we absolutely don't want.
At the end of the day, I'm going to dive a bit more but I might just re-establish the cancellation boolean since I don't think there is a better way... This patch still has a lot of changes we want though. I might need some time to revert some of the changes.

Phew, what a journey 😅

Reply:
Fixed now as discussed in the huddle.